src/caf/toolkit/translation.py

```diff
@@ -480,6 +480,136 @@
+
+    values_col:
+        The name of the column in `matrix` detailing the values to translate.
+
+    translation:
+        A pandas DataFrame defining the weights to use when translating.
+        Needs to contain columns:
+        `translation_from_col`, `translation_to_col`, `translation_factors_col`.
+        When `col_translation` is None, this defines the translation to use
+        for both the rows and columns. When `col_translation` is set, this
+        defines the translation to use for the rows.
+
+    col_translation:
+        A matrix defining the weights to use to translate the columns.
+        Takes an input of the same format as `translation`. When None,
+        `translation` is used as the column translation.
+
+    translation_from_col:
+        The name of the column in `translation` and `col_translation`
+        containing the current index and column values of `matrix`.
+
+    translation_to_col:
+        The name of the column in `translation` and `col_translation`
+        containing the desired output index and column values. This
+        will define the output index and column format.
+
+    translation_factors_col:
+        The name of the column in `translation` and `col_translation`
+        containing the translation weights between `translation_from_col` and
+        `translation_to_col`. Where zone pairs do not exist, they will be
+        infilled with `translate_infill`.
+
+    translation_dtype:
+        The numpy datatype to use to do the translation. If None, then the
+        dtype of `vector` is used. Where such high precision
+        isn't needed, a more memory and time efficient data type can be used.
+
+    check_totals:
+        Whether to check that the input and output matrices sum to the same
+        total.
+
+    index_col_1_out_name:
+        The name to give to `index_col_1_name` on return.
+
+    index_col_2_out_name:
+        The name to give to `index_col_2_name` on return.
+
+    Returns
+    -------
+    translated_matrix:
+        matrix, translated into to_unique_index system.
+
+    Raises
+    ------
+    ValueError:
+        If matrix is not a square array, or if translation any inputs are not
+        the correct format.
+    """
+    # pylint: disable=too-many-arguments, too-many-locals
+    # Init
+    keep_cols = [index_col_1_name, index_col_2_name, values_col]
+    all_cols = matrix.columns.tolist()
+
+    if index_col_1_out_name is None:
+        index_col_1_out_name = index_col_1_name
+    if index_col_2_out_name is None:
+        index_col_2_out_name = index_col_2_name
+
+    # Drop any columns we're not keeping
+    drop_cols = set(all_cols) - set(keep_cols)
+    if len(drop_cols) > 0:
+        warnings.warn(f"Extra columns found in matrix, dropping the following: {drop_cols}")
+    matrix = pd_utils.reindex_cols(df=matrix, columns=keep_cols)
+
+    # Convert to wide to translate
+    wide_mat = pd_utils.long_to_wide_infill(
```

Comment:
Probably one we want to open for discussion in an issue.
At current these functions assume the "index" columns (usually O/D zone ids) aren't the index, but just random columns that we need to define. Because of that, we don't use stack / unstack, but pivot and melt instead.

Will be one of those where we have to change the interface a little bit, but it will likely mean we can coax people into using Pandas a little more efficiently!

Reply:
Issue opened to discuss this in #116